x-pack/filebeat/module/aws/vpcflow/ingest/pipeline.yml

```diff
@@ -128,6 +128,39 @@ processors:
       field: event.kind
       value: event
 
+  - script:
+      lang: painless
+      ignore_failure: true
+      source: |
+        if (ctx?.aws?.vpcflow?.tcp_flags == null)
+          return;
+
+        if (ctx?.aws?.vpcflow?.tcp_flags_array == null) {
+          ArrayList al = new ArrayList();
+          ctx.aws.vpcflow.put("tcp_flags_array", al);
+        }
+
+        def flags = Integer.parseUnsignedInt(ctx.aws.vpcflow.tcp_flags);
+
+        if ((flags & 0x01) != 0) {
+          ctx.aws.vpcflow.tcp_flags_array.add('fin');
+        }
+        if ((flags & 0x02) != 0) {
```

Comment:
I think that if you check for 0x12 first (syn-ack) and then subtract that value out this will be good.

Reply:
> I think that if you check for 0x12 first (syn-ack) and then subtract that value out this will be good.

This could be grumpy old networking guy talking but there is no SYN-ACK flag, you just have the syn and the ack flags set.  So the array that gets populated should contain 'syn' and 'ack'.

```
  /g/k/h/t/n/n/
 /r/c/s/s/y/i/
/u/a/p/r/s/f/
 0 1 0 0 1 0 (0x12 or 18)
```

Which would give you:

```
"aws.vpcflow.tcp_flags_array": ["syn", "ack"]
```

Or we could go with a keyword instead of array where we cat the flags together separated by "-", that would give:

```
"aws.vpcflow.tcp_flags_array": "syn-ack"
```

But I have trouble mixing the two approaches.  For example if all the fields were set would you expect:

```
"aws.vpcflow.tcp_flags_array": ["fin", "syn", "rst", "psh", "ack", "urg"]

or

"aws.vpcflow.tcp_flags_array": ["fin", "syn-ack", "rst", "psh", "urg"]
```

Reply:
I realize that there is no syn-ack flag and that this is a bitmask. 😄 

The reason for the unique handling is that AWS treats ACK special. It only sets the ACK bit when it sees it as part of a SYN,ACK response. If, for example, it sees a plain ACK or PSH,ACK packet in a flow it won't report the ACK bit. So rather than having a plain ACK in the list I was thinking having the SYN-ACK in there to highlight the meaning, but we could just as well report `["SYN", "ACK"]` and have consumers understand the meaning of ACK in this case.

In my experience this is distinctly from how Netflow reports tcp flags. It uses a simple OR based aggregation over the flow. This makes it nearly impossible to determine which side initiated a flow or terminated it. (BTW there's a issue https://github.com/elastic/beats/issues/12858 to do this same thing for netflow.)

Reply:
From a query perspective I don't think it matters too much.  You can query `flags:SYN and not flags.ACK` or  `flags:SYN and flags:ACK`.

Reply:
Yeah, I see that AWS will never report a naked "ACK". 😃 

Looks like AWS will also OR across the aggregation interval (19 for fin, syn & ack), added example of that.

I was also thinking the code could be useful for other data sets like liblogparser.js, which also makes me lean towards just turning the existing flags into names.

That being said, I can live with "syn-ack"

Reply:
I'm good with separate keywords for each bit given that there is no loss of functionality from a query perspective.